openshift/machine-config-operator
# Add bootstrap vs day 2 integration tests based on envtest

**- What I did**

Created a new test suite using a a controller-runtime tool [envtest](https://book.kubebuilder.io/reference/envtest.html), note the new dependency is just so we can use this envtest tool for these tests, no other packages are being imported.

With envtest, the new test suite in `test/e2e-bootstrap/bootstrap_test.go` sets up the controllers from the MCC against a running kubeapi/etcd (provided by envtest).

It then performs the following steps:
* Verifies that the environment is clean (no resources left over from previous tests)
* Creates basic MCPs et al based on data from the `pkg/controller/bootstrap/testdata/bootstrap` folder
* Waits for the controllers to render the config, captures the name for both worker and master pools
* Copies the bootstrap data + additional tests specific manifests into a temp dir
* Runs the bootstrap logic against this temp dir
* Captures the rendered output of the bootstrap tool and compares the names
* If the names of the rendered configs don't match, it prints out both configs for analysis
* Cleanup: Stops the controllers and then deletes all resources created during the test, ready for the next test

The above is captured in the first couple of commits, the remaining commits:
* modify one of the test helpers to allow it to wait for multiple machine config names in the MCP
* Add tests for KubeletConfigs during bootstrap (note these fail currently)
* Add a couple of scripts that allow the relevant pre-test setup to be done (downloading kubeapi/etcd binaries to a temp directory)

**- How to verify it**

You should be able to run `make bootstrap-e2e` on any machine and observe that the first test currently passes, while the remaining three fail. The second should be fixed by #2668 (I have tested this already), and the remaining should be fixed by #2547.

If you already have the "kubebuilder" binaries installed, you can run `make bootstrap-e2e-local` to prevent the test from fetching new binaries.

If this is an acceptable test suite (I appreciate the envtest concept is new to this team), then I will set up a CI check for this test explicitly so that you have a clear signal if there's drift/regressions on this bootstrapping in the future.

Am also happy to jump on a call to do a demo/talk through what this does and how it works. 

Also happy to go on an owners file for this if that would make the team feel more comfortable.
